
Nature makes genomics sequencing for us; we tinker from one string of DNA to other as opposed to starting anew. This brings forth the concept of **Alignment**. Sequences in biology have insertions, deletions, and substitutions, and these are all comparable using the magic of computer science. An interesting paper I read was _A Model of Evolutionary Change in Proteins_ by Dayhoff, Schwartz, and Orcutt, which insinuates that evolution is commutative, it is a vibration around given frequencies. They make a matrix of accepted point mutations, and this has extensions in dynamic programming to find the similarity score between two lengths of DNA/Protein etc.

An elementary resource for doing this is the Needleman-Wunsch algorithm. UCSD has a fantastic visualisation [here](https://bioboot.github.io/bimm143_W20/class-material/nw/) to visualise this. If you have two biological sequences, A and B, and you want to **globally align** them i.e. line them up so you can see which parts match, differ, and where you need gaps, you use this algorithm. Each pair is scored based on a lookup table like BLOSUM. Matching amino acids get a positive score, mismatches get a negative score, and gaps are penalised heavily with a set negative penalty. 

This is essentially a matrix with SeqA as the rows, SeqB as the columns, and the alignment score as the cells. This shows us the conserved regions between the sequences. 
